### PR TITLE
Bind mount `/tmp/.X11-unix` for newer version of WSL2

### LINKED ISSUE #56: [Bug]: Distrod causes X11 to not work under wslg.
### Describe the bug

XWayland under WSLg doesn't work anymore when installing Distrod. It seems like systemd is ignoring the x11.conf that allows the distro to bind-mount /mnt/wslg/.X11-unix to /tmp/.X11-unix

Otherwise, wayland apps run fine.

I am using systemd 251.2-1
Also, I'm using the Windows Store version of WSL.

### Steps to reproduce

1. Install distrod
2. Run a X11 application under WSLg.
3. The application complains that it cannot find :0.

### Expected behavior

For .X11-unix to be properly bind-mounted in /tmp and run XWayland apps.

### Windows version

Microsoft Windows 11 21H2 [Version 10.0.22000.708]

### Linux kernel version

5.10.102.1-microsoft-standard-WSL2

### Distro

ArchWSL 22.3.18.0

### How did you install that distro?

Enabled `distrod` in an existing WSL2 distro

### Logs

[Distrod][DEBUG] distrod-exec: exec_command_in_distro
[Distrod][DEBUG] starting /init from distrod-exec
[Distrod][DEBUG] WSL envs: "WSL_INTEROP" = "/run/WSL/8_interop"
[Distrod][DEBUG] WSL envs: "WSL_DISTRO_NAME" = "ArchLinux"
[Distrod][DEBUG] WSL envs: "WSLENV" = "WT_SESSION::WT_PROFILE_ID"
[Distrod][DEBUG] Container::with_mount source: Some(HostPath("/run/distrod/cmdline")), target: ContainerPath("/proc/cmdline"), fstype: None, flags: MS_BIND, is_file: true
[Distrod][TRACE] mount_distrod_run_files: path: "/opt/distrod/run/systemd"
[Distrod][TRACE] mount_distrod_run_files: path: "/opt/distrod/run/systemd/system"
[Distrod][TRACE] mount_distrod_run_files: path: "/opt/distrod/run/systemd/system/portproxy.service"
[Distrod][DEBUG] Container::with_mount source: Some(HostPath("/opt/distrod/run/systemd/system/portproxy.service")), target: ContainerPath("/run/systemd/system/portproxy.service"), fstype: None, flags: MS_BIND, is_file: true
[Distrod][TRACE] mount_distrod_run_files: path: "/opt/distrod/run/tmpfiles.d"
[Distrod][TRACE] mount_distrod_run_files: path: "/opt/distrod/run/tmpfiles.d/x11.conf"
[Distrod][DEBUG] Container::with_mount source: Some(HostPath("/opt/distrod/run/tmpfiles.d/x11.conf")), target: ContainerPath("/run/tmpfiles.d/x11.conf"), fstype: None, flags: MS_BIND, is_file: true
[Distrod][DEBUG] DistroLauncher::launch
[Distrod][DEBUG] Container::with_mount source: Some(HostPath("/run/distrod/distrod_wsl_env-uid1000")), target: ContainerPath("/run/distrod/distrod_wsl_env-uid1000"), fstype: None, flags: MS_BIND, is_file: true
[Distrod][DEBUG] Spawning the command or the waiter.
[Distrod][DEBUG] Executing a command in the distro.
[Distrod][DEBUG] Failed to ignore signal Sys(EINVAL)
[Distrod][DEBUG] Failed to ignore signal Sys(EINVAL)
[Distrod][DEBUG] Distro::exec_command.
[Distrod][DEBUG] Container::exec_command.
[Distrod][TRACE] mounting source: Some(
    ContainerPath(
        "/run/distrod/cmdline",
    ),
), mount: ContainerMount { source: Some(HostPath("/run/distrod/cmdline")), target: ContainerPath("/proc/cmdline"), fstype: None, flags: MS_BIND, data: None, is_file: true }
[Distrod][TRACE] mounting source: Some(
    ContainerPath(
        "/opt/distrod/run/systemd/system/portproxy.service",
    ),
), mount: ContainerMount { source: Some(HostPath("/opt/distrod/run/systemd/system/portproxy.service")), target: ContainerPath("/run/systemd/system/portproxy.service"), fstype: None, flags: MS_BIND, data: None, is_file: true }
[Distrod][TRACE] mounting source: Some(
    ContainerPath(
        "/opt/distrod/run/tmpfiles.d/x11.conf",
    ),
), mount: ContainerMount { source: Some(HostPath("/opt/distrod/run/tmpfiles.d/x11.conf")), target: ContainerPath("/run/tmpfiles.d/x11.conf"), fstype: None, flags: MS_BIND, data: None, is_file: true }
[Distrod][TRACE] skipping an identical mount: Some(
    ContainerPath(
        "/run/distrod/distrod_wsl_env-uid1000",
    ),
), ContainerMount {
    source: Some(
        HostPath(
            "/run/distrod/distrod_wsl_env-uid1000",
        ),
    ),
    target: ContainerPath(
        "/run/distrod/distrod_wsl_env-uid1000",
    ),
    fstype: None,
    flags: MS_BIND,
    data: None,
    is_file: true,
}
[Distrod][DEBUG] Triple fork done.
[Distrod][DEBUG] dropping privilege. kmsg logging in the child ends here.
[Distrod][DEBUG] The parent of the second of three forks exits.
[Distrod][DEBUG] Spawning the command or the waiter.
[Distrod][DEBUG] Spawning the waiter.
[Distrod][DEBUG] Failed to ignore signal Sys(EINVAL)
[Distrod][DEBUG] Failed to ignore signal Sys(EINVAL)

### additional comment

_No response_

### PR DESCRIPTION
**Pull request by Locietta**
Fix #56.

After [WSL 0.60.0 release](https://github.com/microsoft/WSL/releases/tag/0.60.0), microsoft switched to bind mount the `/tmp/.X11-unix` rather than symlink it as they previously do. This breaks wslg in Distrod for latest WSL on MS Store.

Distrod should bind mount for WSL 0.60.0, and keep the symlink strategy for older versions of WSL.

I've tested the fix on WSL 0.60.0, WSL 0.58.3 and in-box WSL, all of them worked ok.

**EDIT:** Before this PR is merged, you may use opt tar packaged by me for a while 👉 [opt_distrod.tar.gz](https://github.com/nullpo-head/wsl-distrod/files/8985168/opt_distrod.tar.gz)

### CODE REVIEW

**Comment by herrwinfried** [comment]
are you sure you fixed it? it didn't work for me

**Comment by Locietta** [comment]
Hi, @herrwinfried, I'm now using it for myself (WSL0.60.0 and archwsl) and doesn't see any issue🤔

Could you please test [opt_distrod.tar.gz](https://github.com/nullpo-head/wsl-distrod/files/8954538/opt_distrod.tar.gz) to see if it works? 
If it also fails, please paste the output of `ls /run/systemd/system/systemd-tmpfiles-setup.service.d` and `ls /tmp/.X11-unix` to see what's going wrong...

**Comment by herrwinfried** [comment]
> Hi, @herrwinfried, I'm now using it for myself (WSL0.60.0 and archwsl) and doesn't see any issue🤔
> 
> Could you please test [opt_distrod.tar.gz](https://github.com/nullpo-head/wsl-distrod/files/8954538/opt_distrod.tar.gz) to see if it works? If it also fails, please paste the output of `ls /run/systemd/system/systemd-tmpfiles-setup.service.d` and `ls /tmp/.X11-unix` to see what's going wrong...

I'm using openSUSE TW but tmp/.x11-unix looks empty.

**Comment by herrwinfried** [comment]
> Hi, @herrwinfried, I'm now using it for myself (WSL0.60.0 and archwsl) and doesn't see any issue🤔
> 
> Could you please test [opt_distrod.tar.gz](https://github.com/nullpo-head/wsl-distrod/files/8954538/opt_distrod.tar.gz) to see if it works? If it also fails, please paste the output of `ls /run/systemd/system/systemd-tmpfiles-setup.service.d` and `ls /tmp/.X11-unix` to see what's going wrong...

yes this works on the openSUSE TW windows store download you mentioned, thanks.
image

**Comment by Locietta** [comment]
I just notice a bug: when `systemd-tmpfiles --create` is executed, `/tmp/.X11-unix` will be cleaned up.

I've fixed it and updated the opt tar pinned at the beginning.

**Comment by leoleoasd** [comment]
I tried the tarball and it gives the following error everytime i open wsl:
```

**Comment by leoleoasd** [comment]
I'm using latest archlinux.

**Comment by herrwinfried** [comment]
I downloaded the latest edited and it works for me. Could it be a missing step for arch linux. @leoleoasd 
image

**Comment by Locietta** [comment]
> I tried the tarball and it gives the following error everytime i open wsl:
> 
> ```
> C:\Users\Leo>wsl
> -zsh: symbol lookup error: /opt/distrod/lib/libpthread.so.0: undefined symbol: __libc_pthread_init, version GLIBC_PRIVATE
> ```
> 
> Maybe it's something with your compile environment?

It looks like you directly extract the tarball to `/opt`, which might cause the trouble. You are mixing the libs from repo owner (probably package his tarball on ubuntu) and me (on Arch).

Remove the original `/opt/distrod` and re-install the tarball should fix this. @leoleoasd 

And in case you can't open wsl shell, use `wsl -u root -e /opt/distrod/bin/distrod disable` to disable distrod from outside.

**Comment by leoleoasd** [comment]
That worked for me. Thanks!

**Comment by Locietta** [comment]
Seems like CI fails because no `/mnt/wslg/.X11-unix` on test runner for bind mount... while we don't have permission to create it on github runner😥 (see https://github.com/Locietta/wsl-distrod/runs/7183455218?check_suite_focus=true). 

It looks ok when I ran tests on my WSL2 though. Having no idea about how to solve the test failure.

**Comment by nullpo-head** [comment]
Hi @Locietta ! I truly appreciate your contribution! And I'm sorry for my late response. 🙇‍♂️
Overall it looks good to me, but let me check it around tomorrow. I may ask you for some refactoring, or I could do it by myself after merging your contribution. Thanks so much again!

**Review by nullpo-head — APPROVED** [review]
Looks great to me. Thanks so much! I've add a little requests for improvements. I'd appreciate it if you could update it if you can.

File: distrod/distrod/tests/integration_test.rs [hunk 1]
```diff
@@ -326,17 +326,17 @@ fn test_wslg_socket_is_available() {
...
-    let output = ls.output().unwrap();
-    let output = String::from_utf8_lossy(&output.stdout);
-    eprintln!("output of `ls -ld /tmp/.X11-unix`: {}", output);
-    assert!(output.ends_with("-> /mnt/wslg/.X11-unix\n"));
+    // mount x11.conf to symlink
+    let mut test_old = DISTROD_SETUP.new_command();
+    test_old.args(&["exec", "--", "test", "-e", "/run/tmpfiles.d/x11.conf"]);
+    let result_old = test_old.status().unwrap();
+    
+    // mount systemd-tmpfiles-setup.service.d to bind-mount
```
[the remark this answers is not in the document]
Comment (Locietta) [reply distrod/distrod/tests/integration_test.rs]: done

**Comment by dyharlan** [comment]
I tried installing the latest release, and it doesn't seem to fix the issue?

**Comment by Locietta** [comment]
@dyharlan v0.1.7 looks good to me, could you check `ls /run/systemd/system/systemd-tmpfiles-setup.service.d` and `ls /tmp/.X11-unix`?

**Comment by leoleoasd** [comment]
v0.1.7 gives me 
```
<3>init: (12) ERROR: CreateProcessEntryCommon:528: execvpe /opt/distrod/alias/usr/bin/zsh failed 2
<3>init: (12) ERROR: CreateProcessEntryCommon:538: Create process not expected to return
```

Is there something with my installation? I just delete all files inside opt/distrod and replace it with [opt_distrod.tar.gz](https://github.com/nullpo-head/wsl-distrod/releases/download/v0.1.7/opt_distrod.tar.gz)

**Comment by gifnksm** [comment]
In my environment, `/run/tmpfiles.d/x11.conf` was still mounted after upgrading Distrod to v0.17 (even after disabling and enabling Distrod).
I removed everything under `/opt/distrod` and reinstalled Distrod again, and the problem no longer occurs.

**Pull request merged**